File: chapitre7/exercice7.4.py
import sys

def main():
    # 1. Saisie de la taille et "allocation"
    # En Python, la gestion mémoire est automatique, pas de malloc/free [1]
    try:
        n = int(input("Entrez la taille du tableau (n) : "))
    except ValueError:
        print("Erreur : n doit être un entier.")
        return

    # Création d'une liste (équivalent du tableau alloué dynamiquement) [1]
    tab =  * n

    # 2. Remplissage du tableau
    for j in range(n):
        try:
            tab[j] = int(input(f"tab[{j}] = "))
        except ValueError:
            print("Entrée invalide, valeur mise à 0.")

    # 3, 4 & 5. Gestion de l'erreur et point de retour
    # Le bloc try/except remplace ici sigaction, sigsetjmp et siglongjmp [3, 4]
    while True:
        try:
            i_input = input("\nEntrez l'indice i de l'élément à afficher : ")
            if not i_input: 
                break
            i = int(i_input)

            # En C, cette ligne génère un SIGSEGV si i est invalide [1, 2]
            # En Python, cela lève un IndexError
            print(f"Valeur de tab[{i}] = {tab[i]}")
            
            # Si l'affichage réussit, on sort de la boucle (fin du programme)
            break 

        except IndexError:
            # Équivalent du gestionnaire_segv [4, 5]
            print("\n[ERREUR] Erreur de segmentation détectée (i est hors limites) !")
            # La boucle while recommence, simulant le retour au point de sauvegarde
        except ValueError:
            print("Veuillez entrer un nombre entier pour l'indice.")

    print("Fin du programme sans erreur.")

if __name__ == "__main__":
    main()

File: chapitre7/exercice7.4.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <setjmp.h>

// On utilise jmp_buf pour sauvegarder l'état de la pile
// et pouvoir y revenir après un signal
sigjmp_buf point_de_retour;

// Gestionnaire de signal pour SIGSEGV
void gestionnaire_segv(int sig) {
    printf("\n[ERREUR] Erreur de segmentation détectée (i est hors limites) !\n");
    // On retourne au point de sauvegarde défini dans le main
    siglongjmp(point_de_retour, 1);
}

int main() {
    int n, i;
    int *tab;

    // 1. Saisie de la taille et allocation
    printf("Entrez la taille du tableau (n) : ");
    scanf("%d", &n);

    tab = malloc(n * sizeof(int));
    if (tab == NULL) {
        perror("Erreur d'allocation");
        return 1;
    }

    // 2. Remplissage du tableau
    for (int j = 0; j < n; j++) {
        printf("tab[%d] = ", j);
        scanf("%d", &tab[j]);
    }

    // 3. Installation du gestionnaire de signal
    struct sigaction sa;
    sa.sa_handler = gestionnaire_segv;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction(SIGSEGV, &sa, NULL);

    // 4. Point de retour en cas de crash
    // sigsetjmp renvoie 0 la première fois, et 1 quand on revient de siglongjmp
    sigsetjmp(point_de_retour, 1);

    // 5. Saisie de l'indice et affichage
    while (1) {
        printf("\nEntrez l'indice i de l'élément à afficher : ");
        if (scanf("%d", &i) != 1) break;

        // Ici, si i est invalide (ex: 1000000), le système génère un SIGSEGV
        // Le gestionnaire interviendra et nous renverra au sigsetjmp plus haut
        printf("Valeur de tab[%d] = %d\n", i, tab[i]);
        
        // Si on arrive ici sans crash, on peut quitter proprement
        break; 
    }

    free(tab);
    printf("Fin du programme sans erreur.\n");
    return 0;
}